trace/src/main/org/jetbrains/lincheck/trace/Deserialization.kt

```diff
@@ -0,0 +1,771 @@
+    val context: TraceContext = TRACE_CONTEXT
+
+    private val dataStream: SeekableInputStream
+    private val data: SeekableDataInput
+    private val dataBlocks = mutableMapOf<Int, MutableList<DataBlock>>()
+    private val callTracepointChildren = mutableMapOf<Int, Pair<Long, Long>>()
+
+    init {
+        val channel = Files.newByteChannel(Path(dataFileName), StandardOpenOption.READ)
+        dataStream = SeekableChannelBufferedInputStream(channel)
+        data = SeekableDataInput(dataStream)
+
+        try {
+            checkDataHeader(data)
+        } catch (t: Throwable) {
+            data.close()
+            index?.close()
+            throw t
+        }
+    }
+
+    override fun close() {
+        data.close()
+        index?.close()
+    }
+
+    fun readRoots(): List<TRTracePoint> {
+        loadContext()
+
+        val roots = mutableMapOf<Int, TRTracePoint>()
+
+        dataBlocks.forEach {
+            val (threadId, blocks) = it
+            data.seek(blocks.first().physicalStart)
+            val kind = data.readKind()
+            check(kind == ObjectKind.BLOCK_START) { "Thread $threadId block 0 has wrong start: $kind" }
+            val blockId = data.readInt()
+            check(blockId == threadId) { "Thread $threadId block 0 has wrong idt: $blockId" }
+
+            val tracepoints = mutableListOf<TRTracePoint>()
+            loadTracePoints(threadId, Integer.MAX_VALUE, this::readTracePointWithChildAddresses) { _, tracePoint, _
+                -> tracepoints.add(tracePoint)
+            }
+            if (tracepoints.isEmpty()) {
+                System.err.println("Thread $threadId doesn't write any tracepoints")
+            } else {
+                if (tracepoints.size > 1) {
+                    System.err.println("Thread $threadId wrote too many root tracepoints: ${tracepoints.size}")
+                }
+                roots[threadId] = tracepoints.first()
+            }
+        }
+
+        return roots.entries.sortedBy { it.key }.map { (_, tracePoint) -> tracePoint }
+    }
+
+    fun loadAllChildren(parent: TRMethodCallTracePoint) {
+        val positions = callTracepointChildren[parent.eventId]
+            ?: error("TRMethodCallTracePoint ${parent.eventId} is not found in index")
+        data.seek(calculatePhysicalOffset(parent.threadId, positions.first))
+
+        loadTracePoints(parent.threadId, Integer.MAX_VALUE, this::readTracePointWithChildAddresses) { idx, tracePoint, _ ->
+            parent.loadChild(idx, tracePoint)
+        }
+
+        val actualFooterPos = data.position() - 1 // 1 is size of object kind
+        check(actualFooterPos == calculatePhysicalOffset(parent.threadId, positions.second)) {
+            "Input contains broken data: expected Tracepoint Footer for event ${parent.eventId} at position ${positions.second}, got $actualFooterPos"
+        }
+    }
+
+    fun loadChild(parent: TRMethodCallTracePoint, childIdx: Int) = loadChildrenRange(parent, childIdx, 1)
```

Comment:
For `Unit`-return methods, it's better to wrap the body into `{ ... }` instead of using `= ...`. 
Or, at least, add a return type `: Unit` annotation explicitly.